Application/Education/ClassRegister/Diary/Service.php

```diff
@@ -0,0 +1,331 @@
+    {
+
+        $Protocol= '';
+        if(!$withData){
+            $Protocol = (new Setup($this->getStructure()))->setupDatabaseSchema($doSimulation, $UTF8);
+        }
+        if (!$doSimulation && $withData) {
+            (new Data($this->getBinding()))->setupDatabaseContent();
+        }
+        return $Protocol;
+    }
+
+    /**
+     * @param $Id
+     *
+     * @return false|TblDiary
+     */
+    public function getDiaryById($Id)
+    {
+        return (new Data($this->getBinding()))->getDiaryById($Id);
+    }
+
+    /**
+     * @param TblDivision $tblDivision
+     * @param bool $withPredecessorDivision
+     *
+     * @return false|TblDiary[]
+     */
+    public function getDiaryAllByDivision(TblDivision $tblDivision, $withPredecessorDivision = false)
+    {
+        if ($withPredecessorDivision) {
+            $divisionList = array();
+            $resultList = array();
+            $this->getPredecessorDivisionList($tblDivision, $divisionList);
+            /** @var TblDivision $tblDivisionItem */
+            foreach ($divisionList as $tblDivisionItem) {
+                if (($list = $this->getDiaryAllByDivision($tblDivisionItem, false))) {
+                    $resultList = array_merge($resultList, $list);
+                }
+            }
+
+            return $resultList;
+        } else {
+            return (new Data($this->getBinding()))->getDiaryAllByDivision($tblDivision);
+        }
+    }
+
+    /**
+     * @param TblGroup $tblGroup
+     * @param TblYear $tblYear
+     *
+     * @return false|TblGroup[]
+     */
+    public function getDiaryAllByGroup(TblGroup $tblGroup, TblYear $tblYear)
+    {
+        return (new Data($this->getBinding()))->getDiaryAllByGroup($tblGroup, $tblYear);
+    }
+
+    /**
+     * @param TblDivision $tblDivision
+     * @param $resultList
+     */
+    private function getPredecessorDivisionList(TblDivision $tblDivision, &$resultList) {
+        $resultList[$tblDivision->getId()] = $tblDivision;
+        if (($tblDiaryDivisionList = $this->getDiaryDivisionByDivision($tblDivision))) {
+            foreach ($tblDiaryDivisionList as $tblDiaryDivision) {
+                if (($tblPredecessorDivision = $tblDiaryDivision->getServiceTblPredecessorDivision())) {
+                    $this->getPredecessorDivisionList($tblPredecessorDivision, $resultList);
+                }
+            }
+        }
+    }
+
+    /**
+     * @param TblDiary $tblDiary
+     *
+     * @return false|TblDiaryStudent[]
+     */
+    public function getDiaryStudentAllByDiary(TblDiary $tblDiary)
+    {
+        return (new Data($this->getBinding()))->getDiaryStudentAllByDiary($tblDiary);
+    }
+
+    /**
+     * @param $Data
+     * @param TblDivision|null $tblDivision
+     * @param TblGroup|null $tblGroup
+     * @param TblDiary|null $tblDiary
+     *
+     * @return bool|Form
+     */
+    public function checkFormDiary(
+        $Data,
+        TblDivision $tblDivision = null,
+        TblGroup $tblGroup = null,
```

Comment:
Codacy found an issue: [Incorrect spacing between argument "$tblGroup" and equals sign; expected 0 but found 1](https://app.codacy.com/manual/gerdchristian-kunze/SPHERE-Framework/pullRequest?prid=5757982)